### PR TITLE
Trim equal sign suffix from AWS ed25519 encoded fingerprint

### PR DESCRIPTION
**Pull request by Skarlso**
### Description

Closes https://github.com/weaveworks/eksctl/issues/5145

After a long discussion, we determined that AWS adds an extra pad to the fingerprint. Our comparison logic didn't handle that. This is fixing that.

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [x] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by Skarlso** [comment]
Manual test output:

```
 eksctl create nodegroup -f cluster-managed.yaml
2022-04-22 16:51:12 [ℹ]  eksctl version 0.95.0-dev+7cef9efdd.2022-04-22T16:47:11Z
2022-04-22 16:51:12 [ℹ]  using region us-west-2
2022-04-22 16:51:20 [ℹ]  nodegroup "mng-1" will use "" [AmazonLinux2/1.21]
2022-04-22 16:51:20 [ℹ]  using SSH public key "/Users/skarlso/.ssh/id_ed25519.pub" as "eksctl-gb-test-ssh-nodegroup-mng-1-DF3hnpZAPYvtyUJvhKXDz0x/eMWN828yE8ObUpTU06I"
2022-04-22 16:51:20 [ℹ]  nodegroup "mng-2" will use "" [AmazonLinux2/1.21]
2022-04-22 16:51:20 [ℹ]  using SSH public key "/Users/skarlso/.ssh/id_ed25519.pub" as "eksctl-gb-test-ssh-nodegroup-mng-2-DF3hnpZAPYvtyUJvhKXDz0x/eMWN828yE8ObUpTU06I"
2022-04-22 16:51:24 [ℹ]  1 existing nodegroup(s) (mng-1) will be excluded
2022-04-22 16:51:24 [ℹ]  1 nodegroup (mng-2) was included (based on the include/exclude rules)
2022-04-22 16:51:24 [ℹ]  will create a CloudFormation stack for each of 1 managed nodegroups in cluster "gb-test-ssh"
2022-04-22 16:51:24 [ℹ]
2 sequential tasks: { fix cluster compatibility, 1 task: { 1 task: { create managed nodegroup "mng-2" } }
}
2022-04-22 16:51:24 [ℹ]  checking cluster stack for missing resources
2022-04-22 16:51:25 [ℹ]  cluster stack has all required resources
2022-04-22 16:51:26 [ℹ]  building managed nodegroup stack "eksctl-gb-test-ssh-nodegroup-mng-2"
2022-04-22 16:51:26 [ℹ]  deploying stack "eksctl-gb-test-ssh-nodegroup-mng-2"
2022-04-22 16:51:27 [ℹ]  waiting for CloudFormation stack "eksctl-gb-test-ssh-nodegroup-mng-2
```

File: pkg/ssh/client/ssh.go [hunk 1]
```diff
@@ -133,8 +133,11 @@ func importKey(ctx context.Context, keyName, fingerprint string, keyContent *str
 	if existing, err := findKeyInEC2(ctx, ec2API, keyName); err != nil {
 		return err
 	} else if existing != nil {
-		if *existing.KeyFingerprint != fingerprint {
-			return fmt.Errorf("SSH public key %s already exists, but fingerprints don't match (exected: %q, got: %q)", keyName, fingerprint, *existing.KeyFingerprint)
+		// AWS adds an extra `=` to ed25519 sha256 encodings. Normally, that's not there.
```
Comment (nikimanoledaki) [inline pkg/ssh/client/ssh.go:136]: > AWS adds an extra `=` to ed25519 sha256 encodings.

Is there an "and" missing between the two types? The sentence is a bit confusing 🤔

File: pkg/ssh/client/ssh.go [hunk 2]
```diff
@@ -133,8 +133,11 @@ func importKey(ctx context.Context, keyName, fingerprint string, keyContent *str
...
+		// Because it's just base64 padding it's fine to remove and compare without it.
+		trimmedAWSFingerprint := strings.TrimSuffix(*existing.KeyFingerprint, "=")
```
Comment (nikimanoledaki) [inline pkg/ssh/client/ssh.go:138]: Is the padding always going to be up to 1 `=`, not more?

File: pkg/ssh/client/ssh.go (hunk 1 above)
Comment (nikimanoledaki) [inline pkg/ssh/client/ssh.go:136]: > AWS adds an extra `=` to ed25519 sha256 encodings.

Is there an "and" missing between the two types? The sentence is a bit confusing 🤔

Comment (Skarlso) [reply pkg/ssh/client/ssh.go:136]: I'll try to clarify it, sorry. :D

File: pkg/ssh/client/ssh.go (hunk 2 above)
Comment (Skarlso) [reply pkg/ssh/client/ssh.go:138]: Yep :nod:

File: pkg/ssh/client/ssh.go [hunk 3]
```diff
@@ -133,8 +133,12 @@ func importKey(ctx context.Context, keyName, fingerprint string, keyContent *str
...
 		return err
 	} else if existing != nil {
-		if *existing.KeyFingerprint != fingerprint {
-			return fmt.Errorf("SSH public key %s already exists, but fingerprints don't match (exected: %q, got: %q)", keyName, fingerprint, *existing.KeyFingerprint)
+		// AWS uses a slightly different encoding than what Go is doing and ends up with an extra padding.
+		// They are using a combination of base64 and sha256 together. Since sha256 generates 32 bytes,
+		// base64 pads it to 33 with an extra `=`. The padding is safe to remove.
+		trimmedAWSFingerprint := strings.TrimSuffix(*existing.KeyFingerprint, "=")
+		if trimmedAWSFingerprint != fingerprint {
+			return fmt.Errorf("SSH public key %s already exists, but fingerprints don't match (exected: %q, got: %q)", keyName, trimmedAWSFingerprint, *existing.KeyFingerprint)
```
Comment (rtsp) [inline pkg/ssh/client/ssh.go:141]: I think this line parameter swapped. It should be 

```go
return fmt.Errorf("SSH public key %s already exists, but fingerprints don't match (exected: %q, got: %q)", keyName, fingerprint, trimmedAWSFingerprint)
```

BTW, is `exected` should be `expected` ?

Comment (Skarlso) [reply pkg/ssh/client/ssh.go:141]: Yep, you're right in both cases. 😂

**Review by nikimanoledaki — APPROVED** [review]

**Review by nikimanoledaki — APPROVED** [review]

File: pkg/ssh/client/ssh.go [hunk 4]
```diff
@@ -133,8 +133,12 @@ func importKey(ctx context.Context, keyName, fingerprint string, keyContent *str
 	if existing, err := findKeyInEC2(ctx, ec2API, keyName); err != nil {
 		return err
 	} else if existing != nil {
-		if *existing.KeyFingerprint != fingerprint {
-			return fmt.Errorf("SSH public key %s already exists, but fingerprints don't match (exected: %q, got: %q)", keyName, fingerprint, *existing.KeyFingerprint)
+		// AWS uses a slightly different encoding than what Go is doing and ends up with an extra padding.
+		// They are using a combination of base64 and sha256 together. Since sha256 generates 32 bytes,
+		// base64 pads it to 33 with an extra `=`. The padding is safe to remove.
```
Comment (cPu1) [inline pkg/ssh/client/ssh.go:138]: Both AWS and eksctl use the same algorithm to generate the fingerprint for Ed25519 keys, but the Go implementation uses an unpadded base64 encoding (`base64.RawStdEncoding`), which is why the eksctl-generated fingerprint has no padding (=).

**Review by cPu1 — APPROVED** [review]

**Pull request merged**